ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh

```diff
@@ -0,0 +1,201 @@
+                sleep 1m
+                echo -e "wait_time_passed=${wait_time_loop_var} min.\n"
+                if ! ./oc adm upgrade|grep "AdminAckRequired"; then
+                    echo -e "Admin-acks patch still in processing, waiting...\n"
+                else
+                    echo -e "Admin-acks patch finished succ\n"
+                    break
+                fi
+                ((wait_time_loop_var+=1))
+            done
+            [ "${wait_time_loop_var}" -eq 5 ] && echo >&2 "Admin-acks patch not completed in 5 minutes, exiting" && exit 1
+        fi
+    fi
+    set +x
+}
+
+# Upgrade the cluster to target release
+# Parameters: 1- by-digest pull spec, 2- by-tag pull spec
+function upgrade() {
+    set -x
+    local err
+    
+    if ! err="$(./oc adm upgrade --to-image="$1" 2>&1 >/dev/null)" && [[ $err == *"you must pass --allow-explicit-upgrade to continue"* ]]; then
+        echo "Check on oc adm upgrade --to-image missing --allow-explicit-upgrade PASSED"
+    else
+        echo >&2 "Check on oc adm upgrade --to-image missing --allow-explicit-upgrade FAILED: $err, exiting" && exit 1
+    fi
+    local current_image; current_image="$(./oc get clusterversion version -o json|jq -r '.status.history[0].image')"
+    [[ $1 == "${current_image}" ]] && echo >&2 "Upgrade with oc adm upgrade --to-image missing --allow-explicit-upgrade starts but expected NOT start, exiting" && exit 1
+    
+    if [[ "${FORCE_UPDATE}" == "false" ]]; then
+        echo "Upgrading cluster to $1..."
+        if ! ./oc adm upgrade --to-image="$1" --allow-explicit-upgrade; then
+            echo >&2 "Failed to start upgrade, exiting" && exit 1
+        fi
+    else
+        echo "Force upgrading cluster to $2..."
+        if ! ./oc adm upgrade --to-image="$2" --allow-explicit-upgrade --force; then
+            echo >&2 "Failed to start upgrade, exiting" && exit 1
+        fi
+    fi
+
+    set +x
+}
+
+# Monitor the upgrade status
+# Parameter: 1-target release version
+function check_upgrade_status() {
+    local timeout=120
+    local out="", avail="", progress=""
+    while [ ${timeout} -gt 0 ]; do
+        echo "./oc get clusterversion" && ./oc get clusterversion
+        out="$(./oc get clusterversion --no-headers)"
+        avail="$(echo "${out}" | awk '{print $3}')"
+        progress="$(echo "${out}" | awk '{print $4}')"
+        if [[ "${avail}" == "True" && "${progress}" == "False" && "${out}" == *"Cluster version is $1" ]]; then
+            echo -e "Upgrade succ\n\n"
+            break
+        else
+            sleep 5m
+            ((timeout-=5))
+        fi        
+    done
+    # Collect co, node, mcp, cv state for further debug
+    echo "./oc get co" && ./oc get co
+    echo "./oc get node" && ./oc get node
+    echo "./oc get mcp" && ./oc get mcp
+    echo "./oc get clusterversion version -oyaml" && ./oc get clusterversion version -oyaml
+    [ "${timeout}" -le 0 ] && echo >&2 "Upgrade timeout, exiting" && exit 1
+
```

Comment:
Instead of checking only the `clusterversion/version`, will it be better to also take into account the status of `mcp` and `nodes` to conclude a successful upgrade?

Reply:
Yeah, we have the mcp, nodes check in the cluster health check script. If we decide to run cluster health check after each upgrade, it seems I need to add a function to it to replace the current cucushift-upgrade-healthcheck step.

Reply:
Hmm, you mean a function in cucushift-upgrade-toimage-commands.sh or a dedicated bash file for cluster-health-check. I thought there are many common functions in this file, which could be re-used for other `upgrade commands`.

Reply:
>there are many common functions in this file which could be re-used for other upgrade commands.

Right. I implemented all of them in cucushift-upgrade-toimage-commands.sh instead of extracting to a bash file respectively because I'm not sure if a bash file can be called in cucushift-upgrade-toimage-commands.sh. I don't see this kind of usage. Let me check with DPTP.

Reply:
health check can be implemented as a dedicated step

Reply:
@jhou1 If implementing health check as a step, how can I run this step in the cucushift-upgrade-toimage-commands.sh after each round upgrade?

Reply:
Unfortunately, I didn't find that possible...

Reply:
It's not possible to use a static file in commands.sh today. 
1. In current structure (looping multiple upgrades in a step), if we want to do cluster health check after each upgrade round, we have to implement it as a function in commands.sh. The original heath check script contains too many checks, can we select a few core checks, like co, node, pod, mcp check, to assemble them into commands.sh?
2. Any possibility to change the structure, one upgrade has a step? With this way I have no idea how can we pass the target release to a upgrade step.

Any ideas/suggestions?

Reply:
I'm not sure I understand the question in 2. Seem like you want to recursively upgrade the cluster until it is the target version and assemble them in one step. This way you have to implement the health check as a function. 

Prow config does not allow duplicate steps, so a health check step can't be invoked multiple times in a workflow.

Reply:
Thanks Jianwei. Then option 2 is not possible.
Let's talk about option 1, any suggestion on the health check function would be appreciated @jiajliu

Reply:
Can we have a talk on this? I remember we have similar issue on eus upgrade. I suggest to split them to each round upgrade/check. But i'm not sure how jianwei deal with it now.